src/vsr/grid.zig

```diff
@@ -391,7 +394,12 @@ pub fn GridType(comptime Storage: type) type {
 
         pub fn cancel(grid: *Grid, callback: *const fn (*Grid) void) void {
             // grid.open() is cancellable the same way that read_block()/write_block() are.
-            assert(grid.callback == .none or grid.callback == .open);
+            switch (grid.callback) {
+                .none => {},
+                .open => {},
+                .checkpoint => unreachable,
+                .cancel => unreachable,
+            }
```

Comment:
Any specific reason to avoid 

```zig
.none, .open => {}
.checkpoint, .cancel => unreachable,
```

grouping here?

Reply:
:thinking: I think we have one extra invariant here that we aren't actually checking perhaps? For states which are not cancelable (so, checkpoint and cancel), we must also ensure that we don't _read_ any blocks, as that could deadlock. 

In particular, we shouldn't be reading blocks during checkpoint. So, `assert(grid.callback != .cancel);` in `read_block` should also check `callback != chcekpoint`, and, really, it should be the same as current assert. 

so maybe add `gird.callback.can_cancel()` helper for assertions here and on the read path?

Reply:
In `grid.read_block` we already have:

```
            switch (callback) {
                .from_local_storage => {
                    maybe(grid.callback == .checkpoint);
                    ...
                },
                .from_local_or_global_storage => {
                    assert(grid.callback != .checkpoint);
                    ...
                },
            }
```

Which I think is correct -- reads to satisfy grid repair requests are safe.

Reply:
> Any specific reason to avoid

Just personal preference -- I like all of the variants aligned, so if I was grouping them I would do

```
.none,
.open,
=> {},
.checkpoint,
.cancel,
=> unreachable,
```

But in this case that is longer than just repeating the rhs.